File: benchmark/src/inverse.rb
def swap_rows(matrix, row1, row2, n)
    for i in 0...n
        temp = matrix[row1][i]
        matrix[row1][i] = matrix[row2][i]
        matrix[row2][i] = temp
    end
end

def gauss_jordan(matrix, inverse, n)
    for i in 0...n
        if matrix[i][i] == 0
            for j in (i + 1)...n
                if matrix[j][i] != 0
                    swap_rows(matrix, i, j, n)
                    swap_rows(inverse, i, j, n)
                    break
                end
            end
        end
        pivot = matrix[i][i]
        for j in 0...n
            matrix[i][j] /= pivot
            inverse[i][j] /= pivot
        end
        for j in 0...n
            if j != i
                factor = matrix[j][i]
                for k in 0...n
                    matrix[j][k] -= factor * matrix[i][k]
                    inverse[j][k] -= factor * inverse[i][k]
                end
            end
        end
    end
end

input_file = File.open("input/input.txt", "r")
output_file = File.open("output/ruby.txt", "w")

n = input_file.gets.to_i

matrix = Array.new(n) { Array.new(n) }
inverse = Array.new(n) { Array.new(n, 0) }

for i in 0...n
    matrix[i] = input_file.gets.chomp.split.map(&:to_f)
    inverse[i][i] = 1.0
end

gauss_jordan(matrix, inverse, n)

for i in 0...n
    output_file.puts inverse[i].join(" ")
end

input_file.close
output_file.close

File: benchmark/src/inverse.pl
use strict;
use warnings;

sub swap_rows {
    my ($matrix, $row1, $row2, $n) = @_;
    for my $i (0 .. $n - 1) {
        my $temp = $matrix->[$row1]->[$i];
        $matrix->[$row1]->[$i] = $matrix->[$row2]->[$i];
        $matrix->[$row2]->[$i] = $temp;
    }
}

sub gauss_jordan {
    my ($matrix, $inverse, $n) = @_;
    for my $i (0 .. $n - 1) {
        if ($matrix->[$i]->[$i] == 0) {
            for my $j ($i + 1 .. $n - 1) {
                if ($matrix->[$j]->[$i] != 0) {
                    swap_rows($matrix, $i, $j, $n);
                    swap_rows($inverse, $i, $j, $n);
                    last;
                }
            }
        }
        my $pivot = $matrix->[$i]->[$i];
        for my $j (0 .. $n - 1) {
            $matrix->[$i]->[$j] /= $pivot;
            $inverse->[$i]->[$j] /= $pivot;
        }
        for my $j (0 .. $n - 1) {
            if ($j != $i) {
                my $factor = $matrix->[$j]->[$i];
                for my $k (0 .. $n - 1) {
                    $matrix->[$j]->[$k] -= $factor * $matrix->[$i]->[$k];
                    $inverse->[$j]->[$k] -= $factor * $inverse->[$i]->[$k];
                }
            }
        }
    }
}

open my $input_file, '<', 'input/input.txt' or die "Error opening file!";
open my $output_file, '>', 'output/perl.txt' or die "Error opening file!";

my $n = <$input_file>;
chomp($n);

my @matrix;
my @inverse;

for my $i (0 .. $n - 1) {
    my @row = split ' ', <$input_file>;
    push @matrix, \@row;
    push @inverse, [(map {0} 1 .. $n)];
    $inverse[$i][$i] = 1;
}

gauss_jordan(\@matrix, \@inverse, $n);

for my $i (0 .. $n - 1) {
    print $output_file join(' ', @{$inverse[$i]}), "\n";
}

close $input_file;
close $output_file;
